src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +295,119 @@ def process_input(conf):
+            os.link(cfn, dst)
+
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
+
+        if conf.retrieve_sample:
+            _download_sample_data(conf, upa, metafile)
+
+
+def _download_sample_data(conf: Conf, upa: str, metafile: str) -> None:
+    # retrieve sample data from sample service and save to file
+    # additionally, retrieve node data from the sample data and save it to a file
+
+    with open(metafile, "r", encoding="utf8") as json_file:
+        meta = json.load(json_file)
+
+    upa_dir = Path(metafile).parent
+    sample_file_name = f"{upa}.{loader_common_names.SAMPLE_FILE_EXT}"
+    sample_file = os.path.join(upa_dir, sample_file_name)
+
+    if (loader_common_names.SAMPLE_FILE_KEY in meta and
+            meta[loader_common_names.SAMPLE_FILE_KEY] == sample_file_name and
+            os.path.isfile(sample_file)):
+        print(f"Skip downloading sample for {upa} as it already exists")
+        return
 
-        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
-        # upa file is downloaded to cfn
-        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+    # retrieve data links associated with upa
+    links_ret = conf.ss.get_data_links_from_data({"upa": upa.replace("_", "/")})
 
-        # each upa in output_dir as a seperate directory
-        dstd = os.path.join(conf.pth, upa)
-        os.makedirs(dstd, exist_ok=True)
+    data_links = links_ret['links']
+    if not data_links and conf.ignore_no_sample_error:
+        print(f"No sample data links found for {upa}")
+        return
 
-        dst = os.path.join(dstd, f"{upa}.fa")
-        # Hard link .fa file from job_dir to output_dir in WS
-        os.link(cfn, dst)
+    # there should only be one data link for each upa
+    if len(data_links) != 1:
+        raise ValueError(f"Expected 1 data link for {upa}, got {len(data_links)}")
 
-        metafile = os.path.join(dstd, f"{upa}.meta")
-        # save meta file with relevant object_info
-        with open(metafile, "w", encoding="utf8") as json_file:
-            json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+    # retrieve sample data and save to file
+    sample_id = data_links[0]['id']
+    sample_ret = conf.ss.get_sample_via_data({"upa": upa.replace("_", "/"),
+                                              "id": sample_id,
+                                              "version": data_links[0]["version"]})
 
-        print("Completed %s" % (upa))
+    with open(sample_file, "w", encoding="utf8") as json_file:
+        json.dump(sample_ret, json_file, indent=2)
+
+    meta[loader_common_names.SAMPLE_FILE_KEY] = sample_file_name
```

Comment:
Can we add the munged sample file here as well? And probably best to write the metadata after both files are saved.

Reply:
👍

Reply:
:+1: